nomad/structs/csi.go

```diff
@@ -406,3 +434,142 @@ type CSIVolumeGetResponse struct {
 	Volume *CSIVolume
 	QueryMeta
 }
+
+// CSIPlugin bundles job and info context for the plugin for clients
+type CSIPlugin struct {
+	ID        string
+	Driver    string
+	Type      CSIPluginType
+	Namespace string // FIXME all jobs in the same namespace?
+	Jobs      map[string]map[string]*Job
+
+	ControllerHealthy int
+	Controllers       map[string]*CSIInfo
+	NodeHealthy       int
+	Nodes             map[string]*CSIInfo
+
+	CreateIndex uint64
+	ModifyIndex uint64
+}
+
+func NewCSIPlugin(id, driver string, index uint64) *CSIPlugin {
+	return &CSIPlugin{
+		ID:          id,
+		Driver:      driver,
+		Jobs:        map[string]map[string]*Job{},
+		Controllers: map[string]*CSIInfo{},
+		Nodes:       map[string]*CSIInfo{},
+		CreateIndex: index,
+		ModifyIndex: index,
+	}
+}
+
+func (p *CSIPlugin) AddPlugin(nodeID string, info *CSIInfo, index uint64) {
+	if info.ControllerInfo != nil {
+		prev, ok := p.Controllers[nodeID]
+		if ok && prev.Healthy {
+			p.ControllerHealthy -= 1
+		}
+		p.Controllers[nodeID] = info
+		if info.Healthy {
+			p.ControllerHealthy += 1
+		}
+	}
+
+	if info.NodeInfo != nil {
+		prev, ok := p.Nodes[nodeID]
+		if ok && prev.Healthy {
+			p.NodeHealthy -= 1
+		}
+		p.Nodes[nodeID] = info
+		if info.Healthy {
+			p.NodeHealthy += 1
+		}
+	}
+
+	p.ModifyIndex = index
+}
+
+func (p *CSIPlugin) DeletePlugins(nodeID string, index uint64) {
```

Comment:
If I understand correctly, when we call this we then check if the plugin's list of nodes is empty and only then delete it from the state store. Should this be `DeleteNodes`?

Reply:
More of a question than a thing to fix: it looks like `CSIPlugin` gets mutated only within a transaction and then the struct gets treated as immutable after that, so we don't have to worry about concurrent changes to `NodeHealthy` and `ControllerHealthy`. Do we have any mechanisms to enforce that kind of thing other than "be careful"?

Reply:
`DeleteNodes` sounds good, and it seems like we've got two parts to the transaction thing: memdb does enforce a single writer at a time, but these structs should be changed copy-on-write to avoid errors in threads that have an open read-only transaction and are closed over the structs

Reply:
The plugins are now copy on write, and denormalization is aware that other structs are too.